### DISCUSSION TITLE
Method not found: void BasicPublish() exception when using latest MonoSDK and Xamarin.iOS

### DISCUSSION THREAD

**Discussion by bnikollov** [rabbitmq/rabbitmq-dotnet-client#1033] [General]
UPDATED!

We are using the RabbitMQ (6.2.1 NuGet) in a Xamarin .Net Standard 2.1 project. After we build the project with the latest Mono SDK and Xamarin.iOS SDK we got issues. 
**The first issue** is that the client (iOS application) is successfully connected, but the queues could not be consumed by the client. We do not get any errors/exceptions regarding that.  
**The second issue** is that for some reason the BasicPublish method is missing and we are not able to pubish the new created messages from the Xamarin.iOS app to the RabbitMQ queues. Here is the exception:

"DeviceModel":"Apple iPhone","OS":"iOS 14.4.1","message":"Exception","ExceptionMessage":"Method not found: void RabbitMQ.Client.IModelExensions.BasicPublish(RabbitMQ.Client.IModel,string,string,RabbitMQ.Client.IBasicProperties,System.ReadOnlyMemory`1<byte>)","StackTrace":" at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[TResult].Start[TStateMachine] (TStateMachine& stateMachine) [0x0002c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:471 \n at Vapp.Services.Network.QueueApi.WebMessageClient.SendMessageAsync (Vapp.Models.Network.QueueApi.Outgoing.OutgoingWebMessage message) [0x0000c] in <8b277fb515eb472fb32c2af9d425a279>:0 \n 

We thought it could a Linker issue but the error is still there if turn off the Linker at all.

If we build the client application with Mono 6.12.0.113, Xamarin.iOS 14.8.0.3 RabbitMQ works as expected, but other issues are rising (non related to RabbitMQ). 
If we use higher versions the issue appears.

Thanks in advance! :)

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
None of the lines in the trace point to this client. You haven't specified what .NET client version is used but at least in master, [IModelExtensions](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/master/projects/RabbitMQ.Client/client/api/IModelExtensions.cs) now uses `CachedString` instead of `string`. This is [not the case in `6.x`](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/6.x/projects/RabbitMQ.Client/client/api/IModelExtensions.cs), however.

This change was introduced in https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/990. I know nothing about how/why Xamarin uses or distributes our client but it's up to them to adapt to the changes if they depend on master and not the latest `6.x.` release.

**Reply by bnikollov** [reply]
Hi, Michael! Thank you for the fast response! We use the latest stable version (6.2.1) that is available as NuGet package. Xamarin does not distributes your client! We use it in our cross-platform project. I will gather more info regarding the issue that we are facing and post it here.

**Reply by bnikollov** [reply]
I have updated the initial issue, please check it again. If you need other information, please let me know. Thanks!

**Reply by michaelklishin** [reply]
Like I said, in master the signature of the method has changed to use a `CachedString`, see https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/990. I would not recommend using master until 7.0 ships, and it will contain breaking changes like this one, so you will have to adjust your code.

I cannot comment on the "other issues" and it's not clear to me why you are trying to use a "more recent version" and what that version is. We haven't shipped any releases after 6.2.1.

**Comment by bnikollov** [comment]
_"Like I said, in master the signature of the method has changed to use a CachedString, see #990."_

- I don't get that how this change affects the version that we are using (6.2.1)?

![](https://user-images.githubusercontent.com/18344640/113981179-0abe9580-9850-11eb-9c94-69296f37a344.png)


_"I would not recommend using master until 7.0 ships, and it will contain breaking changes like this one, so you will have to adjust your code."_

- We will have that in mind and prepare for the upcoming changes, thanks. But we are currently depending on version 6.2.1 and it would be great if you could support us on finding and probably fixing the issue with the missing method.

_"I cannot comment on the "other issues""_

- That is why I have just mentioned them as other issues (not related RabbitMQ).

_"...and it's not clear to me why you are trying to use a "more recent version" and what that version is. We haven't shipped any releases after 6.2.1."_

- We are not trying to use more recent version of RabbitMQ.Client, but we should always use the latest versions of MonoSDK and Xamarin.iOS in order to be able to build and compile iOS application that runs properly on the latest version of iOS. For now there are no serious bugs / issues when we use the older versions, but after a few updates we could be blocked from releasing a new version of our application.

**Reply by karimElJed** [reply]
@michaelklishin We did a support case with Microsoft and found out that there is an issue within netstandard 2.1. See 
https://developercommunity.visualstudio.com/t/xamarinios-application-compiles-but-has-problems-a/1396737

The current version of RabbitMQ.Client is built for netstandard 2.0. Do you have any plans to upgrade to netstandard 2.1? This would fix the issue with the missing method.

**Reply by lukebakken** [reply]
No, `netstandard2.1` is a dead-end.

**Comment by lukebakken** [comment]
Hello everyone who has commented here -

If you have time to test version 7 alphas, it would be really appreciated

https://www.nuget.org/packages/RabbitMQ.Client/7.0.0-alpha.6

I'm hoping to get the final release done this month, but only if there is enough community testing (with reports back!)
